blogs/2022/09/15/dev-container-features.md

```diff
@@ -0,0 +1,147 @@
+
+```
++-- feature
+|    +-- devcontainer-feature.json
+|    +-- install.sh
+|    +-- (other files)
+
+```
+
+`install.sh`: The installation entry point script – it’s conceptually added as a layer of the image’s Dockerfile and executed during build-time. This entry point script can install tools such as languages (ruby) and tools (GitHub CLI).
+
+`devcontainer-feature.json`: This contains metadata about the Feature, a set of options that can be passed to a Feature’s install script during installation, as well as "pieces" of devcontainer.json that will be merged into the final dev container. For example, if any Feature indicates `privileged`: true in their config, the entire dev container will be started with the `--privileged` flag.
+
+Features can be authored in a variety of languages, the most straightforward being shell scripts. If a Feature is authored in a different language, information about it should be included in the metadata so that users can make an informed choice about it.
+
+> **Note:** While the install.sh will run Features in any language, if you wrote a Feature in an interpreted language that wasn't present on the build machine, the code would fail to execute.
+
+> You should be sure that Features you release publicly check and install dependencies in addition to the Feature.
+
+> Furthermore, public Features are very likely to be used from both arm64 or x86_64 machines - so be sure to adapt to this when possible.
+
+You may review `devcontainer-feature.json` properties [in the spec](https://containers.dev/implementors/features/#devcontainer-feature-json-properties), along with public examples in the [devcontainers/features](https://github.com/devcontainers/features) repo.
+
+Now that we’ve seen how to create a Feature, how can I distribute it to others?
+
+## Distribution
+
+Features are distributed as tarballs. The tarball contains the entire contents of the Feature sub-directory, including the `devcontainer-feature.json`, `install.sh`, and any other files in the directory.
+
+The [Open Container Initiative](https://opencontainers.org/), aka OCI, defines industry standards for containers and container resources. We treat Features as OCI Artifacts and use the concept of an [OCI Registry](https://containers.dev/implementors/features-distribution/#oci-registry) to distribute Features.
+
+The [Features template repository](https://github.com/devcontainers/feature-template) mentioned above includes a [GitHub Actions workflow](https://github.com/marketplace/actions/dev-container-publish) to automate the publishing process. It packages each feature into a tarball and publishes the assets as an OCI artifact to GHCR. Trigger the `release.yaml` workflow from the template repository by selecting it on the left of the repository’s Actions tab on GitHub. This will publish each feature to GHCR under the `<owner>/<repo>` namespace. A Feature is only republished when the version property in its `devcontainer-feature.json` is updated.
```

Comment:
Should we just talk about a single feature?
"It package the feature into a tarball...."
"This GitHub Action will publish the feature to GHCR...."

Reply:
I like this idea, updated.